### PR TITLE
updated/refactored code

### PR DESCRIPTION
**Pull request by Gamerkuipers**
**Changes**

- changed `sofort/digital` to `sofort/physical`
- changed `testMode` check to short if else 
- abstract functions `getTitle`, `getDescription` to 2 functions in abstract.php instead of in the gateway classes
- consts for `apiresponses `
- made the empty option for ideal issuer `hidden` so you can't select it after selecting an issuer
- function `debug` instead of wrapping the code with an if added a gaurd to return if debug mode is off
- function `woocommerce3` removed the if because `version_compare()` return a `bool` if you use the `optional operator `[php.net - version_compare()](https://www.php.net/manual/en/function.version-compare.php)
- updated a `spelling mistake`
- added `check` for `invalid apikey`
- added default `empty array` to `ideal_issuers` in `ideal gateway` (so it doesnt php error dump in the ideal description)
- refactored/moved file `gateways/abstract.php` to `gateway.php` & made it `non abstract `

### CODE REVIEW

File: paypro-gateways-woocommerce/languages/paypro-gateways-woocommerce-nl_NL.po [hunk 1]
```diff
@@ -60,8 +60,8 @@ msgid "%s payment in process (%s)"
 msgstr "%s betaling in gang (%s)"
 
 #: includes/paypro/wc/gateway/abstract.php:153
```
Comment (iownthegame) [inline paypro-gateways-woocommerce/languages/paypro-gateways-woocommerce-nl_NL.po:62]: I see the filename is changed, maybe we want to change this comment as well? Same for L63

File: paypro-gateways-woocommerce/includes/paypro/wc/gateway.php [hunk 2]
```diff
@@ -1,18 +1,24 @@
...
 
-    protected $default_description;
+    protected $default_title;
 
     protected $default_logo;
 
     protected $issuer;
+    protected $gateway_title = '';
+    protected $gateway_description = '';
+    public $has_fields = FALSE;
```
Comment (iownthegame) [inline paypro-gateways-woocommerce/includes/paypro/wc/gateway.php:17]: Does this need to be `public`?

File: paypro-gateways-woocommerce/includes/paypro/wc/gateway.php [hunk 3]
```diff
@@ -1,18 +1,24 @@
...
+	protected const PAYPROAPIRES_NOTSUBSCRIPTED = "Not subscribed to money transfer service";
 
-    protected $default_description;
+    protected $default_title;
 
     protected $default_logo;
 
     protected $issuer;
+    protected $gateway_title = '';
+    protected $gateway_description = '';
```
Comment (iownthegame) [inline paypro-gateways-woocommerce/includes/paypro/wc/gateway.php:16]: Shouldn't it be `__('', 'paypro-gateways-woocommerce')`? or it doesn't matter

File: paypro-gateways-woocommerce/includes/paypro/wc/gateway.php [hunk 4]
```diff
@@ -1,18 +1,24 @@
...
+class PayPro_WC_Gateway extends WC_Payment_Gateway
 {
-    protected $default_title;
+	/**
+	 * panther restapi responses
+	 */
+	protected const PAYPROAPIRES_NOTSUBSCRIPTED = "Not subscribed to money transfer service";
 
-    protected $default_description;
+    protected $default_title;
```
Comment (iownthegame) [inline paypro-gateways-woocommerce/includes/paypro/wc/gateway.php:10]: Why do we keep this but remove `$default_description`? what about `$default_logo`

File: paypro-gateways-woocommerce/includes/paypro/wc/gateway/ideal.php [hunk 5]
```diff
@@ -32,9 +26,9 @@ public function payment_fields()
         $selected_issuer = $this->getSelectedIssuer();
 
         $html = '<select name="' . PayPro_WC_Plugin::PLUGIN_ID . '_issuer_' . $this->id . '">';
-        $html .= '<option value=""></option>';
+        $html .= '<option value="" hidden></option>';
```
Comment (iownthegame) [inline paypro-gateways-woocommerce/includes/paypro/wc/gateway/ideal.php:29]: I am not sure if I get this, could you make a screenshot for it? is it always hidden?

Comment (Gamerkuipers) [reply paypro-gateways-woocommerce/includes/paypro/wc/gateway/ideal.php:29]: it will still be selected as `default` bc its the first option in the list but you wont see the `empty` option in the list of options

Comment (Gamerkuipers) [reply paypro-gateways-woocommerce/includes/paypro/wc/gateway/ideal.php:29]: image

Comment (Gamerkuipers) [reply paypro-gateways-woocommerce/includes/paypro/wc/gateway/ideal.php:29]: image

File: paypro-gateways-woocommerce/includes/paypro/wc/gateway.php (hunk 2 above)
Comment (Gamerkuipers) [reply paypro-gateways-woocommerce/includes/paypro/wc/gateway.php:17]: image
wordpress will also throw a critical error

image

File: paypro-gateways-woocommerce/includes/paypro/wc/gateway.php (hunk 3 above)
Comment (Gamerkuipers) [reply paypro-gateways-woocommerce/includes/paypro/wc/gateway.php:16]: in my eyes it doesnt matter since its an empty string and it would just translate to an empty string and it is not allowed to use those expressions there. (i dont know if there is a way to have them als default but if it is requested i can add it to the constructer and set them `$this->gateway_title = __('', 'paypro-gateways-woocommerce')` (same for description)

image

also a critical error on wordpress
image

File: paypro-gateways-woocommerce/languages/paypro-gateways-woocommerce-nl_NL.po (hunk 1 above)
Comment (Gamerkuipers) [reply paypro-gateways-woocommerce/languages/paypro-gateways-woocommerce-nl_NL.po:62]: will do :)

File: paypro-gateways-woocommerce/includes/paypro/wc/gateway.php (hunk 4 above)
Comment (Gamerkuipers) [reply paypro-gateways-woocommerce/includes/paypro/wc/gateway.php:10]: no clue why they were actually there to begin with i removed them now and will be pushed in the next push

File: paypro-gateways-woocommerce/includes/paypro/wc/gateway.php [hunk 6]
```diff
@@ -1,18 +1,20 @@
...
 {
-    protected $default_title;
-
-    protected $default_description;
-
-    protected $default_logo;
+	/**
+	 * panther restapi responses
+	 */
+	protected const PAYPROAPIRES_NOTSUBSCRIPTED = "Not subscribed to money transfer service";
```
Comment (iownthegame) [inline paypro-gateways-woocommerce/includes/paypro/wc/gateway.php:8]: The indentation is different with the lines below

Comment (iownthegame) [reply paypro-gateways-woocommerce/includes/paypro/wc/gateway.php:8]: Could be make it `PAYPRO_API_RES_NOT_SUBSCRIPTED` for more readable? Also I am thinking about `SUBSCRIBED` instead of `SUBSCRIPTED`

File: paypro-gateways-woocommerce/includes/paypro/wc/gateway.php (hunk 2 above)
Comment (iownthegame) [reply paypro-gateways-woocommerce/includes/paypro/wc/gateway.php:13]: I see, it is a `public` attribute https://woocommerce.wp-a2z.org/oik_api/wc_payment_gatewayhas_fields/

File: paypro-gateways-woocommerce/includes/paypro/wc/gateway.php (hunk 6 above)
Comment (Gamerkuipers) [reply paypro-gateways-woocommerce/includes/paypro/wc/gateway.php]: updated, also updated it for `apikey` invalid in `PayproAPIHelper`

File: paypro-gateways-woocommerce/includes/paypro/wc/gateway.php [hunk 7]
```diff
@@ -1,18 +1,20 @@
...
 {
-    protected $default_title;
-
-    protected $default_description;
-
-    protected $default_logo;
+	/**
+	 * panther restapi responses
+	 */
+	protected const PAYPRO_API_RES_NOT_SUBSCRIBED = "Not subscribed to money transfer service";
```
Comment (iownthegame) [inline paypro-gateways-woocommerce/includes/paypro/wc/gateway.php:8]: This message seems to belong to PayProApiHelper, could we define it there?

Comment (Gamerkuipers) [reply paypro-gateways-woocommerce/includes/paypro/wc/gateway.php]: done

**Review by iownthegame — APPROVED** [review]

File: paypro-gateways-woocommerce/includes/paypro/api/PayProApiHelper.php [hunk 8]
```diff
@@ -41,10 +47,15 @@ public function getSaleStatus($payment_hash)
 
     private function execute()
     {
-        if($this->testMode) $this->api->setParam('test_mode', 'true'); else $this->api->setParam('test_mode', 'false');
+        $this->api->setParam('test_mode', ($this->testMode ? 'true' : 'false'));
 
         $result = $this->api->execute();
 
+        if($result['return'] == self::PAYPRO_API_RES_APIKEY_INVALID) {
+            $result['errors'] = 'true';
```
Comment (paypro-leon) [inline paypro-gateways-woocommerce/includes/paypro/api/PayProApiHelper.php:55]: Do we also want to set a message here?

File: paypro-gateways-woocommerce/includes/paypro/api/PayProApiHelper.php [hunk 9]
```diff
@@ -19,7 +25,7 @@ public function getIdealIssuers()
     {
         $this->api->setCommand('get_all_pay_methods');
         $result = $this->execute();
-
+
```
Comment (paypro-leon) [inline paypro-gateways-woocommerce/includes/paypro/api/PayProApiHelper.php:28]: This change is not necessary.

File: paypro-gateways-woocommerce/includes/paypro/wc/gateway.php [hunk 10]
```diff
@@ -1,18 +1,15 @@
 <?php if(!defined('ABSPATH')) exit; // Exit if accessed directly
 
-abstract class PayPro_WC_Gateway_Abstract extends WC_Payment_Gateway
+class PayPro_WC_Gateway extends WC_Payment_Gateway
```
Comment (paypro-leon) [inline paypro-gateways-woocommerce/includes/paypro/wc/gateway.php:3]: Should we not keep this `abstract`? We do not want this to be instantiated by itself.

File: paypro-gateways-woocommerce/includes/paypro/wc/gateway.php [hunk 11]
```diff
@@ -146,13 +143,15 @@ public function process_payment($order_id)
         if($result['errors'])
         {
             PayPro_WC_Plugin::debug($this->id . ': Failed to create payment for order ' . $order_id . ' - Message: ' .$result['message']);
-
+
```
Comment (paypro-leon) [inline paypro-gateways-woocommerce/includes/paypro/wc/gateway.php:146]: Unnecessary change

File: paypro-gateways-woocommerce/includes/paypro/wc/gateway.php [hunk 12]
```diff
@@ -1,18 +1,15 @@
...
+class PayPro_WC_Gateway extends WC_Payment_Gateway
 {
-    protected $default_title;
-
-    protected $default_description;
-
-    protected $default_logo;
-
     protected $issuer;
+    protected $gateway_title = '';
```
Comment (paypro-leon) [inline paypro-gateways-woocommerce/includes/paypro/wc/gateway.php:6]: We could also make these constants which can be changed by the child class.

File: paypro-gateways-woocommerce/includes/paypro/wc/gateway/ideal.php [hunk 13]
```diff
@@ -32,9 +26,9 @@ public function payment_fields()
...
 
-        foreach($ideal_issuers['issuers'] as $issuer)
+        foreach($ideal_issuers['issuers'] ?? [] as $issuer)
```
Comment (paypro-leon) [inline paypro-gateways-woocommerce/includes/paypro/wc/gateway/ideal.php:31]: We currently still support PHP versions lower than 7.0 which means we cannot use `??` operator.

File: paypro-gateways-woocommerce/includes/paypro/wc/plugin.php [hunk 14]
```diff
@@ -4,7 +4,7 @@ class PayPro_WC_Plugin
 {
     const PLUGIN_ID = 'paypro-gateways-woocommerce';
     const PLUGIN_TITLE = 'PayPro Gateways - WooCommerce';
-    const PLUGIN_VERSION = '1.3.3';
+    const PLUGIN_VERSION = '1.3.4';
```
Comment (paypro-leon) [inline paypro-gateways-woocommerce/includes/paypro/wc/plugin.php:7]: We should make a separate PR for this.

File: paypro-gateways-woocommerce/paypro-gateways-woocommerce.php [hunk 15]
```diff
@@ -4,7 +4,7 @@
  * Plugin Name: PayPro Gateways - WooCommerce
  * Plugin URI: https://www.paypro.nl/
  * Description: With this plugin you easily add all PayPro payment gateways to your WooCommerce webshop.
- * Version: 1.3.3
+ * Version: 1.3.4
```
Comment (paypro-leon) [inline paypro-gateways-woocommerce/paypro-gateways-woocommerce.php:7]: We should make a separate PR for this.

**Review by paypro-leon — CHANGES_REQUESTED** [review]

File: paypro-gateways-woocommerce/includes/paypro/api/PayProApiHelper.php (hunk 8 above)
Comment (Gamerkuipers) [reply paypro-gateways-woocommerce/includes/paypro/api/PayProApiHelper.php:55]: There is already a message given by the api itself `API key not valid` which will then also go thru the new error handling which will end up as `There was an issue processing your payment.` in the frontend

File: paypro-gateways-woocommerce/includes/paypro/wc/gateway/ideal.php (hunk 13 above)
Comment (Gamerkuipers) [reply paypro-gateways-woocommerce/includes/paypro/wc/gateway/ideal.php:31]: will be updated

File: paypro-gateways-woocommerce/includes/paypro/wc/gateway.php (hunk 10 above)
Comment (Gamerkuipers) [reply paypro-gateways-woocommerce/includes/paypro/wc/gateway.php:3]: Could you explain a bit more, i don't get it

File: paypro-gateways-woocommerce/includes/paypro/wc/gateway.php (hunk 12 above)
Comment (Gamerkuipers) [reply paypro-gateways-woocommerce/includes/paypro/wc/gateway.php:6]: `issuer`, `gateway_title` and `gateway_description`?
